File: why_LeetCode/Day05/C++/3Sum Closest.cpp
'''
œ»≈≈–Ú£¨»ª∫Û◊Û”“º–±∆£¨∏¥‘”∂»O(n2)

'''

class Solution {
public:
    int threeSumClosest(vector<int>& nums, int target) {
        int result=0;
        int min_gap=INT_MAX;
        sort(nums.begin(),nums.end());
        
        auto begin=nums.begin(),last=nums.end()-1;
        for(auto i=begin;i!=prev(last,1);i++)
        {
            auto j=next(i);
            auto k=last;
            while(j<k)
            {
                const int sum=*(i)+*(j)+*(k);
                const int gap=abs(sum-target);
                if(gap<min_gap)
                {
                    result=sum;
                    min_gap=gap;
                }
                if(sum<target) 
                    j++;
                else if(sum>target)
                    k--;
                else
                    return sum;
            }
        }
        return result;
    }
};

File: why_LeetCode/Day05/Python/3Sum Closest.py
class Solution:
    def threeSumClosest(self, nums: List[int], target: int) -> int:
        result=0
        min_gap=sys.maxsize
        nums.sort()
        
        for i in range(len(nums)-2):
            j=i+1;
            k=len(nums)-1
            
            while j<k:
                total=nums[i]+nums[j]+nums[k]
                gap=abs(total-target)
                if gap<min_gap:
                    min_gap=gap
                    result=total
                if total<target:
                    j+=1
                elif total>target:
                    k-=1
                else:
                    return total
        return result